base/utf16.jl

```diff
@@ -20,100 +45,240 @@ function endof(s::UTF16String)
+    # get number of bytes to allocate
+    len, flags, num4byte, num3byte, num2byte = unsafe_checkstring(dat, 1, len>>>1)
+    flags == 0 && @inbounds return UTF8String(copy!(Vector{UInt8}(len), dat))
+    return encode_to_utf8(UInt16, dat, len + num2byte + num3byte*2 + num4byte*3)
+end
+
+"
+Converts a `UTF16String` to a `UTF8String`
+
+### Returns:
+*   `UTF8String`
+
+### Throws:
+*   `UnicodeError`
+"
+function convert(::Type{UTF8String}, str::UTF16String)
+    dat = str.data
+    len = sizeof(dat) >>> 1
+    # handle zero length string quickly
+    len <= 1 && return empty_utf8
+    # get number of bytes to allocate
+    len, flags, num4byte, num3byte, num2byte = unsafe_checkstring(dat, 1, len-1)
+    flags == 0 && @inbounds return UTF8String(copy!(Vector{UInt8}(len), 1, dat, 1, len))
+    return encode_to_utf8(UInt16, dat, len + num2byte + num3byte*2 + num4byte*3)
+end
+
+"
+Converts an already validated vector of `UInt16` or `UInt32` to a `UTF8String`
+
+### Input Arguments:
+* `dat` Vector of code units (`UInt16` or `UInt32`), explicit `\0` is not converted
+* `len` length of output in bytes
+
+### Returns:
+* `UTF8String`
+"
+function encode_to_utf8{T<:Union{UInt16, UInt32}}(::Type{T}, dat, len)
+    buf = Vector{UInt8}(len)
+    out = 0
+    pos = 0
+    @inbounds while out < len
+        ch::UInt32 = dat[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle 0x80-0x7ff
+        elseif ch < 0x800
+            buf[out += 1] = 0xc0 | (ch >>> 6)
+            buf[out += 1] = 0x80 | (ch & 0x3f)
+        # Handle 0x10000-0x10ffff (if input is UInt32)
+        elseif ch > 0xffff # this is only for T == UInt32, should not be generated for UInt16
+            output_utf8_4byte!(buf, out, ch)
+            out += 4
+        # Handle surrogate pairs
+        elseif is_surrogate_codeunit(ch)
+            output_utf8_4byte!(buf, out, get_supplementary(ch, dat[pos += 1]))
+            out += 4
+        # Handle 0x800-0xd7ff, 0xe000-0xffff UCS-2 characters
         else
-            i += 1
+            buf[out += 1] = 0xe0 | ((ch >>> 12) & 0x3f)
+            buf[out += 1] = 0x80 | ((ch >>> 6) & 0x3f)
+            buf[out += 1] = 0x80 | (ch & 0x3f)
         end
     end
-    return i > n || !utf16_is_surrogate(data[i])
+    UTF8String(buf)
 end
 
-function convert(::Type{UTF16String}, data::AbstractVector{UInt16})
-    !isvalid(UTF16String, data) && throw(UnicodeError(UTF_ERR_INVALID_16,0,0))
-    len = length(data)
-    d = Array(UInt16, len + 1)
-    d[end] = 0 # NULL terminate
-    UTF16String(copy!(d,1, data,1, len))
+function convert(::Type{UTF16String}, str::ASCIIString)
+    dat = str.data
+    @inbounds return fast_utf_copy(UTF16String, UInt16, length(dat), dat, true)
 end
 
+convert(::Type{Vector{UInt16}}, str::UTF16String) = str.data
+convert(::Type{Array{UInt16}},  str::UTF16String) = str.data
+
+convert(::Type{UTF16String}, str::UTF16String)    = str
```

Comment:
@nalimilan Let me explain about the 3 types of functions:

  * The constructors: `ASCIIString`, `UTF8String`, `UTF16String`, `UTF32String`.
    These are totally unsafe, and really should not be used in user code.  They do no validation
    (except for the stupid, inconsistent, insistence of `UTF16String` and `UTF32String` to have a `0`
    word after the real data).  They build a string that directly uses the `Vector` passed in (see #11391, where there was another inconsistency that I fixed earlier).
     In another PR, I'd like to make the unsafe behavior harder to reach, with maybe a Val argument that must be true, to get the unsafe current behavior, which is currently needed for the conversions)

  * The convert methods.  These are very inconsistent, and even after this set of PRs, there are still more bugs that need to be addressed (a couple of them are #11501, #11502, but there are still a lot more to work through :disappointed: )  As I said before, the `convert{T}(::Type{T}, x::T) = x` really should be doing validation/conversion, but that could be expensive, unless we make sure that all strings are validated (which will help performance, because other operations don't need to worry about bad data, and improve security as well).   The conversion methods are all I'm trying to fix in these PRs.

 *  The `ascii`, `utf8`, `utf16`, and `utf32` convenience functions.
     These are, for the most part, simply wrappers around doing a `convert(::T, ....)`, except that they also add loading from a `Ptr{}`, both `\0` terminated and counted.

@tkelman I hadn't done the `AbstractVector{UInt16}` method before, because I hadn't touched any of the `AbstractArray` and `AbstractVector` methods that were already there...  I've gone ahead and fixed it now, since you raised the question, I just hope Stefan doesn't come down on me for doing more than one thing in a PR! :grinning: